File: usaco/Solution/Section 3.1/humble.pas
{
ID: a_zaky01
PROG: humble
LANG: PASCAL
}

const
  max=2147483647;

var
  k,n,i,j,min:longint;
  p:array[1..100] of longint;
  h:array[0..100000] of int64;
  temp:array[1..100] of longint;

begin
  assign(input,'humble.in'); reset(input);
  assign(output,'humble.out'); rewrite(output);
  readln(k,n);
  for i:=1 to k do read(p[i]);
  h[0]:=1;
  //fillchar(temp,sizeof(temp),0);
  for i:=1 to n do
    begin
      min:=max;
      for j:=1 to k do
        if p[j]*h[temp[j]]<min then min:=p[j]*h[temp[j]];
      h[i]:=min;
      for j:=1 to k do
        if p[j]*h[temp[j]]=min then inc(temp[j]);
    end;
  writeln(h[n]);
  close(input);
  close(output);
end.

File: usaco/Solution/Section 3.1/humble.cpp
/*
ID: a_zaky01
PROG: humble
LANG: C++
*/

#include <stdio.h>
#include <stdlib.h>

int k,n,p[99],temp[99];
long long h[100001];
FILE *fin,*fout;

int main(){
    fin=fopen("humble.in","r");
    fout=fopen("humble.out","w");
    fscanf(fin,"%d %d",&k,&n);
    for (int i=0; i<k; i++)
        fscanf(fin,"%d",&p[i]);
    h[0]=1;
    for (int i=1; i<=n; i++){
        h[i]=p[0]*h[temp[0]];
        for (int j=1; j<k; j++)
            if (p[j]*h[temp[j]]<h[i])
               h[i]=p[j]*h[temp[j]];
        for (int j=0; j<k; j++)
            if (p[j]*h[temp[j]]==h[i])
               temp[j]++;
               }
    fprintf(fout,"%lld\n",h[n]);
    //system ("pause");
    return 0;
}       
